src/hooks/useButton.ts

```diff
@@ -0,0 +1,18 @@
+import { useState } from 'react'
+
+export const useButton = () => {
+  const [isLoading, setIsLoading] = useState(false)
+
+  const handleClick = async <ExpectedResult>(callback: () => Promise<ExpectedResult | null>) => {
```

Comment:
it is safer to use `useCallback`. Otherwise, the hook that is dependent on this function will be triggered at every rendering
```suggestion
  const handleClick = useCallback(async <ExpectedResult>(callback: () => Promise<ExpectedResult | null>) => {
```

Reply:
Plus I think handle click should return a function, not a promise.

Reply:
> Plus I think handle click should return a function, not a promise.

I was thinking of passing requests in the `handleClick` function like so

```js
const res = await handleClick<JoinClassResults>(async () => {
      return await request<JoinClassParams, JoinClassResults>(`joinClass`, {
        code,
      })
    })
```

and disabling buttons until the promise is fulfilled in which case the response is assigned to the variable and can be used like so:)

```js
if (res?.cid) {
      dispatch(enroll({ name: res.name, cid: res.cid, code }))
    }
```

if no response is expected we can pass in `<void>` as `<ExpectedResult>`

Reply:
hmm... I think your current implementation seems correct. I don't remember what I meant.